Create LimitXYAccelJerk.py

GregValiant opened the pull request:
New post processor to limit the X and/or Y Accel and Jerk for bed-slinger printers.
This is a response to Cura Feature Request #10596 and  #8797
# Description

This adds an M201 line to change the Max Accel within the printer.  That effectively limits any M204 lines in the gcode.
If altering the Jerk is enabled then the changes are line-by-line within the gcode as there is no limit on jerk.  Both the Accel and Jerk are reset at the end of the gcode.

This fixes... OR This improves... -->
Tall thin prints on bed-slinger printers can break loose from the build plate so changing the Accel at a layer is very helpful.

<!-- Please delete options that are not relevant. -->

- [ X] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

- [ X] Test A  Cura 4 and 5

**Test Configuration**:
* Operating System:  Win 10 and Marlin

# Checklist:

- [X ] My code follows the style guidelines of this project as described in [UltiMaker Meta](https://github.com/Ultimaker/Meta) and [Cura QML best practices](https://github.com/Ultimaker/Cura/wiki/QML-Best-Practices)
- [ X] I have read the [Contribution guide](https://github.com/Ultimaker/Cura/blob/main/contributing.md) 
- [X ] I have commented my code, particularly in hard-to-understand areas
- [X ] I have uploaded any files required to test this change

rburema commented on plugins/PostProcessingPlugin/scripts/LimitXYAccelJerk.py:
```diff
@@ -0,0 +1,273 @@
+        mycura = CuraApplication.getInstance().getGlobalContainerStack()
+        extruder = mycura.extruderList
+        constant_change = not bool(self.getSettingValueByKey("gradient_change"))
+        accel_print_enabled = bool(extruder[0].getProperty("acceleration_enabled", "value"))
```
See my remarks about 'single extrusion' in the main review.

rburema reviewed (changes requested):
Hi Greg,

Sorry it took a while to get back. Thanks again for all the work & enthusiasm.

Overall this looks acceptable (minus the very small code-style remark), with the exception that this won't work at all for multiple extruders.

Even while there may be precious few 'bed-slinger' printers out there with more than one (which brings up another point -- are we sure the Limit X/Y Accel name will only attract those users?) -- I think the script should at least not be 'silent' about this; that is, it just applies the settings for one extruder to the other.

The simplest solution I'd accept is just to make it clear in the comments & append the scripts name with '... (single extrusion only)' or something similar.

GregValiant commented:
Thanks Remco.
I see where "Mullti-extruders" could end up with "un-intended consequences" if the extruders had different Accel values.

- Added a warning so the user can make a choice of whether to let the post run anyway.
- Added a warning (and exit) to UM users regarding fixed beds.  For other fixed bed printers I don't know that there is a way to explain.  It can't hurt anything, just won't have any effect except to lengthen the print time.
- Fixed a bug that could occur if a user tried to raise the accel limits rather than lowering them.
- Changed variable names to lower case.

(I'm having issues with my new laptop.  Things are iffy here at the moment.  My alternate computer is my Win7 CAD computer that has very limited capability for anything other than CAD.)

rburema commented:
Hi Greg,

Thanks for the update. Unfortunately, something seems to have gone wrong -- though, if I only keep the 'stashed changes' and manually delete the 'upstream changes' (as defined in the last commit) it seems to go alright (as far as I can test with my current setup anyway).

Is UltiMaker really the only company with fixed beds? What about delta-type printers? I think it's alright if we'd just limit the waring to the multi-extruder one, and drop the other warning. After all, it still will technically work then, which is different from the confusion of the singe- versus multi-extruder ones.

GregValiant commented:
I dropped my laptop off for repair yesterday.  When I get it back I'll straighten out the "this version, that version" thing.  This Win7 machine is 32 bit and it has an older version of Python so I'm hamstrung right now.

In regards to the warnings, it could be simpler.  Excluding printers with "Elliptical" beds might be an option.
The post won't do any harm if a printer has a fixed bed...it just won't do any good either.

GregValiant commented on plugins/PostProcessingPlugin/scripts/LimitXYAccelJerk.py:
```diff
@@ -0,0 +1,273 @@
+        x_jerk = int(self.getSettingValueByKey("X_jerk"))  
+        y_jerk = int(self.getSettingValueByKey("Y_jerk"))
+        # Put the strings together
+        M201_limit_new = "M201 X" + x_accel + " Y" + y_accel
```
Done

GregValiant commented:
Alright.  I think I have this updated  with the requested changes.

rburema commented:
Looks good to me! Except perhaps that the 'except' clause with 'all' should probably be 'pass'.

Also, I see you use local variables like `accel_travel` instead of class variables (which would then be `self._accel_travel` or some such -- the underscore isn't strictly nescesary, but would serve to indicate a private variable). The advantages of class variables is that they persist over the object, so you can use them in the other function in the object (for that reason, they are often instantiated in the init method). -- The way you set it up now will also work though, since you just retrieve them again in the execute method.

I'm merging this to another branch first, as there are conflicts with git(hub) for some reason.
